controllers/dependencies.go

```diff
@@ -0,0 +1,167 @@
+	"k8s.io/client-go/restmapper"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	OPERATOR_PIPELINES_REPO = "https://github.com/redhat-openshift-ecosystem/operator-pipelines.git"
+	REPO_CLONE_PATH         = "/tmp/operator-pipelines/"
+	PIPELINE_MANIFESTS_PATH = "ansible/roles/operator-pipeline/templates/openshift/pipelines"
+	TASKS_MANIFESTS_PATH    = "ansible/roles/operator-pipeline/templates/openshift/tasks"
+)
+
+func (r *OperatorPipelineReconciler) reconcilePipelineDependencies(meta metav1.ObjectMeta) error {
+
+	// Cloning operator-pipelines project to retrieve pipelines and tasks
+	// yaml manifests that need to be applied beforehand
+	// ref: https://github.com/redhat-openshift-ecosystem/certification-releases/blob/main/4.9/ga/ci-pipeline.md#step-6---install-the-certification-pipeline-and-dependencies-into-the-cluster
+
+	_, err := git.PlainClone(REPO_CLONE_PATH, false, &git.CloneOptions{
+		URL:      OPERATOR_PIPELINES_REPO,
+		Progress: os.Stdout,
+	})
+	if err != nil {
+		log.Log.Info("Couldn't clone the repository for operator-pipelines.")
+		return err
+	}
+
+	// Reading pipeline manifests and applying to cluster
+	root := REPO_CLONE_PATH + PIPELINE_MANIFESTS_PATH
+	err = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() {
+			if errors := r.applyManifests(path, meta.Namespace); errors != nil {
+				return errors
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		log.Log.Info("Couldn't iterate over operator-pipelines yaml manifest files")
+		return err
+	}
+
+	// Reading tasks manifests and applying it to cluster
+	root = REPO_CLONE_PATH + TASKS_MANIFESTS_PATH
+	err = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() {
+			if errors := r.applyManifests(path, meta.Namespace); errors != nil {
+				return errors
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		log.Log.Info("Couldn't iterate over operator-pipelines yaml manifest files")
+		return err
+	}
+
+	// Removing cloned project
+	if err = os.RemoveAll(PIPELINE_MANIFESTS_PATH); err != nil {
+		log.Log.Info("Couldn't remove operator-pipelines directory")
+		return err
+	}
+	return nil
+}
+
+func (r *OperatorPipelineReconciler) applyManifests(fileName string, Namespace string) error {
+
+	b, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		log.Log.Info("Couldn't read manifest file", "File:", fileName)
+		return err
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Log.Info("Couldn't get in cluster config.")
+		return err
+	}
+
+	c, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Log.Info("Couldn't initialize kubernetes client from config.")
+		return err
+	}
+
+	dd, err := dynamic.NewForConfig(config)
+	if err != nil {
+		log.Log.Info("Couldn't initialize dynamic k8s client from config.")
+		return err
+	}
+
+	// Decoding yaml files to resource objects and apply to cluster
+	decoder := yamlutil.NewYAMLOrJSONDecoder(bytes.NewReader(b), 100)
```

Comment:
I was having some fun today to resolve the same issue, and I think using yaml.Unmarshal is cleaner. This is the snippet of my code to read a csv file:
```
	yfile, err := os.ReadFile("csv.yaml")
	if err != nil {
		log.Error(err)
	}

	var data operatorv1alpha1.ClusterServiceVersion
	err = yaml.Unmarshal(yfile, &data)
```
I checked the pipeline and task folders and we need to handle two CRDs (Task and Pipeline). I think this could be simplified. Please take a look at how we manage a similar situation in preflight (but it is the opposite! we try to dump go struct into yaml manifest). https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/1a2ef8a3829ea911356cd41a0171130d568379cd/certification/internal/policy/operator/deployable_by_olm.go#L378

Please let me know if we need to discuss further.

Reply:
As it is it's working. Are we aiming for refactoring here? Wouldn't it be interesting having another iteration with another issue to simplify it? I would suggest that we unblock other tasks that depends on this step first and take the code refactoring and new approaches on another issue. Glad to jump in and make the changes there. What do you think?

Reply:
Yes refactoring.. I understand this is working, but I think it is too complex to follow and some of these lines could be removed. I'd personally like to push a code that is easy to maintain the first time. The refactoring is usually more complicated as you need to have/add unit tests to ensure the refactored code is not breaking, and we don't have unit tests yet.

BTW, have you seen this?
https://github.com/kubernetes/client-go/issues/193#issuecomment-343138889

Reply:
Hi Samira, nope. I didn't see this one. But just realized that going the generic approach with client-go was not necessary there. So I went on the specific approach which is actually way easier. I just flipped to controller runtime client and that's much better now I think. Please check it out.